js/src/forum/components/DiscussionList.js

```diff
@@ -11,56 +11,38 @@ import Placeholder from '../../common/components/Placeholder';
  *
  * - `params` A map of parameters used to construct a refined parameter object
  *   to send along in the API request to get discussion results.
+ * - `state` A DiscussionListState object that represents the discussion lists's state.
  */
 export default class DiscussionList extends Component {
   init() {
-    /**
-     * Whether or not discussion results are loading.
-     *
-     * @type {Boolean}
-     */
-    this.loading = true;
-
-    /**
-     * Whether or not there are more results that can be loaded.
-     *
-     * @type {Boolean}
-     */
-    this.moreResults = false;
-
-    /**
-     * The discussions in the discussion list.
-     *
-     * @type {Discussion[]}
-     */
-    this.discussions = [];
-
-    this.refresh();
+    this.state = this.props.state;
   }
 
   view() {
-    const params = this.props.params;
+    const state = this.state;
+
+    const params = state.getParams();
     let loading;
 
-    if (this.loading) {
+    if (state.isLoading()) {
       loading = LoadingIndicator.component();
-    } else if (this.moreResults) {
+    } else if (state.moreResults) {
       loading = Button.component({
         children: app.translator.trans('core.forum.discussion_list.load_more_button'),
         className: 'Button',
-        onclick: this.loadMore.bind(this),
+        onclick: state.loadMore.bind(state),
       });
     }
 
-    if (this.discussions.length === 0 && !this.loading) {
+    if (!state.hasDiscussions() && !state.isLoading()) {
       const text = app.translator.trans('core.forum.discussion_list.empty_text');
       return <div className="DiscussionList">{Placeholder.component({ text })}</div>;
     }
 
     return (
-      <div className={'DiscussionList' + (this.props.params.q ? ' DiscussionList--searchResults' : '')}>
+      <div className={'DiscussionList' + (state.params.q ? ' DiscussionList--searchResults' : '')}>
```

Comment:
This one looks like we're breaking encapsulation. (We're using `getParams()` elsewhere. But really, I'd rather see a method that has a name that explains the semantics of what we're doing here - after all, the `q` param already has special meaning within the state class, IIRC.)

Reply:
Agreed, and done: this is clearer